Improve LayoutLM

NielsRogge opened the pull request:
# What does this PR do?

- [x] Improve documentation of `LayoutLM`, explaining how people can normalize bounding boxes before passing them to the model, add links to the various datasets on which the model achieves state-of-the-art results, add code examples in the documentation for the various models
- [x] Add notebook to the list of community notebooks showcasing how to fine-tune `LayoutLMForTokenClassification` on the [FUNSD](https://guillaumejaume.github.io/FUNSD/) dataset (on which the model achieves SOTA results)
- [x] Add integration tests, which confirm that the model outputs the same tensors as the original implementation on the same input data
- [x] Add `LayoutLMForSequenceClassification`, which makes it possible to fine-tune LayoutLM for document image classification tasks (such as the [RVL-CLIP dataset](https://www.cs.cmu.edu/~aharley/rvl-cdip/)), extra tests included.

Fixes the following issues:
- #9228 
- #9097 
- #8866 
- #8524

## Who can review?

@LysandreJik, @patrickvonplaten, @sgugger

patrickvonplaten commented on docs/source/model_doc/layoutlm.rst:
```diff
@@ -19,26 +19,63 @@ Overview
-- LayoutLM has an extra input called :obj:`bbox`, which is the bounding boxes of the input tokens.
-- The :obj:`bbox` requires the data that on 0-1000 scale, which means you should normalize the bounding box before
-  passing them into model.
+- LayoutLM has an extra input compared to BERT called :obj:`bbox`, which are the bounding boxes (i.e. 2D-positions) of
```
```suggestion
- In addition to `input_ids`, :meth:`~transformer.LayoutLMModel.forward` also expects the input :obj:`bbox`, which are the bounding boxes (i.e. 2D-positions) of
```

patrickvonplaten commented on docs/source/model_doc/layoutlm.rst:
```diff
@@ -19,26 +19,63 @@ Overview
+
+   image = Image.open("name_of_your_document - can be a png file, pdf, etc.")
+
+   width, height = image.size
```
that's very nice

patrickvonplaten commented on notebooks/README.md:
```diff
@@ -76,3 +76,4 @@ Pull Request so it can be included under the Community notebooks.
 |[Fine-tuning TAPAS on Sequential Question Answering (SQA)](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb) | How to fine-tune *TapasForQuestionAnswering* with a *tapas-base* checkpoint on the Sequential Question Answering (SQA) dataset | [Niels Rogge](https://github.com/nielsrogge) | [![Open In Colab](https://colab.resear ...
 |[Evaluating TAPAS on Table Fact Checking (TabFact)](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/TAPAS/Evaluating_TAPAS_on_the_Tabfact_test_set.ipynb) | How to evaluate a fine-tuned *TapasForSequenceClassification* with a *tapas-base-finetuned-tabfact* checkpoint using a combination of the 🤗 datasets and 🤗 transformers libraries | [Niels Rogge](https://github.com/nielsrogge)  ...
 |[Fine-tuning mBART for translation](https://colab.research.google.com/github/vasudevgupta7/huggingface-tutorials/blob/main/translation_training.ipynb) | How to fine-tune mBART using Seq2SeqTrainer for Hindi to English translation | [Vasudev Gupta](https://github.com/vasudevgupta7) | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/git ...
+|[Fine-tuning LayoutLM on FUNSD (a form understanding dataset)](https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLM/Fine_tuning_LayoutLMForTokenClassification_on_FUNSD.ipynb) | How to fine-tune *LayoutLMForTokenClassification* on the FUNSD dataset | [Niels Rogge](https://github.com/nielsrogge) | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](htt ...
```
Awesome!

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -638,8 +640,10 @@ def _init_weights(self, module):
+            Bounding boxes of each input sequence tokens. Selected in the range ``[0,
+            config.max_2d_position_embeddings-1]``. Each bounding box should be a normalized version in (x0, y0, x1,
+            y1) format, where (x0, y0) corresponds to the position of the upper left in the bounding box, and (x1, y1)
+            represents the position of the lower right. See Overview section for normalization.
```
Could we link the "Overview section for normalization" here directly? Or maybe put this in the glossary. It would be very nice if it's absolutely trivial for the user to know how to preprocess his image to get the bbox

NielsRogge replied:
Linked it to the Overview section

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -872,7 +878,52 @@ def forward(
+            >>> tokens = []
+            >>> token_boxes = []
+            >>> for word, box in zip(words, normalized_word_boxes):
+            ...    word_tokens = tokenizer.tokenize(word)
```
think it should be `>>>` here as well instead of `...` no @sgugger ?

LysandreJik replied:
For multi-line, it should be `...` after the first line, otherwise the python console interprets each line as standalone.

sgugger replied:
Not sure the indent is right though, it looks like there is a space missing.

NielsRogge replied:
I copied the code example into a notebook and it runs (to make sure people can directly use it). So not sure if there's anything wrong :)

LysandreJik replied:
Yeah, there's a space missing for it to be 4-spaces indented, but Python is quite permissive regarding the indentation as long as it's constant. Since it's the only line indented, Python doesn't complain!

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -872,7 +878,52 @@ def forward(
+            >>> tokens = []
+            >>> token_boxes = []
+            >>> for word, box in zip(words, normalized_word_boxes):
+            ...    word_tokens = tokenizer.tokenize(word)
```
can't we do directly `word_ids = tokenizer(word)` here? Don't really want to incentivize the user too much to use `tokenizer` and `convert_tokens_to_ids`

NielsRogge replied:
Updated this :)

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -872,7 +878,52 @@ def forward(
+            >>> token_boxes = [[0, 0, 0, 0]] + token_boxes + [[1000, 1000, 1000, 1000]]
+
+            >>> input_ids = torch.tensor(tokenizer.convert_tokens_to_ids(tokens)).unsqueeze(0) # batch size of 1
+            >>> bbox = torch.tensor(token_boxes).unsqueeze(0) # batch size of 1
```
(nit)
```suggestion
            >>> bbox = torch.tensor([token_boxes])
```

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -913,10 +964,140 @@ def forward(
+class LayoutLMForSequenceClassification(LayoutLMPreTrainedModel):
+    config_class = LayoutLMConfig
+    pretrained_model_archive_map = LAYOUTLM_PRETRAINED_MODEL_ARCHIVE_LIST
+    base_model_prefix = "layoutlm"
```
this is already inherited from `LayoutLMPreTrainedModel` I think so no need to add it here (probably the same applies for other `LayoutLMFor...`)

patrickvonplaten commented on tests/test_modeling_layoutlm.py:
```diff
@@ -209,36 +235,266 @@ def test_for_masked_lm(self):
+
+
+@require_torch
+class LayoutLMModelIntegrationTest(unittest.TestCase):
```
very nice!

patrickvonplaten commented on tests/test_modeling_layoutlm.py:
```diff
@@ -209,36 +235,266 @@ def test_for_masked_lm(self):
+            device=torch_device,
+        )
+
+        self.assertTrue(torch.allclose(outputs.last_hidden_state[0, :3, :3], expected_slice, atol=TOLERANCE))
```
(nit) I like it better to directly write `atol=1e-3` here -> this way one doesn't have to look for `TOLERANCE`

patrickvonplaten commented on tests/test_modeling_layoutlm.py:
```diff
@@ -209,36 +235,266 @@ def test_for_masked_lm(self):
-    def test_LayoutLM_backward_pass_reduces_loss(self):
-        """Test loss/gradients same as reference implementation, for example."""
-        pass
+        self.assertTrue(torch.allclose(outputs.loss, expected_loss, atol=0.1))
```
`atol=1e-3` would not pass here?

patrickvonplaten commented on tests/test_modeling_layoutlm.py:
```diff
@@ -17,19 +17,26 @@
 from transformers import is_torch_available
-from transformers.file_utils import cached_property
-from transformers.testing_utils import require_torch, require_torch_gpu, slow, torch_device
+from transformers.testing_utils import require_torch, slow, torch_device
```
Very nice tests!

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -958,6 +1134,51 @@ def forward(
+
+            >>> tokens = []
+            >>> token_boxes = []
+            >>> for word, box in zip(words, normalized_word_boxes):
```
Same as previous, think it should be `>>>` below. Also, it would be great if we could just use `tokenizer(...)` instead of `tokenize(...)` and `convert_tokens_to_ids`. Is the problem that we need to know the length of each sample and this doesn't work if the input is batched? If it's the case I think it'd be better to compute `token_boxes` separately and try to just use one function call for `input_ids`

sgugger replied:
Can't see with the comment, but there might be a space missing there too ;-)

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -913,10 +964,140 @@ def forward(
+            >>> tokens = []
+            >>> token_boxes = []
+            >>> for word, box in zip(words, normalized_word_boxes):
+            ...    word_tokens = tokenizer.tokenize(word)
```
same here regarding example

sgugger replied:
Same space missing in the indent

patrickvonplaten reviewed (commented):
Very nice PR! Thanks so much for taking care of this. The notebook looks great as well. Left a couple of comments. If possible it would be awesome if we could make the example a bit more concise (*e.g.* to just use `tokenizer(...)` instead of `tokenize(...)` and `conevrt_tokens_to_ids(...)`.

sgugger commented on docs/source/model_doc/layoutlm.rst:
```diff
@@ -19,26 +19,63 @@ Overview
+- LayoutLM has an extra input compared to BERT called :obj:`bbox`, which are the bounding boxes (i.e. 2D-positions) of
+  the input tokens. These can be obtained using an external OCR engine such as Google's `Tesseract
+  <https://pypi.org/project/pytesseract/>`__. Each bounding box should be in (x0, y0, x1, y1) format, where (x0, y0)
+  corresponds to the position of the upper left in the bounding box, and (x1, y1) represents the position of the lower
+  right. Note that one first needs to normalize the bounding boxes to be on a 0-1000 scale. To normalize, you can use
```
```suggestion
  corresponds to the position of the upper left corner in the bounding box, and (x1, y1) represents the position of the lower
  right corner. Note that one first needs to normalize the bounding boxes to be on a 0-1000 scale. To normalize, you can use
```

sgugger commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -638,8 +640,10 @@ def _init_weights(self, module):
-            config.max_2d_position_embeddings-1]``.
+            Bounding boxes of each input sequence tokens. Selected in the range ``[0,
+            config.max_2d_position_embeddings-1]``. Each bounding box should be a normalized version in (x0, y0, x1,
+            y1) format, where (x0, y0) corresponds to the position of the upper left in the bounding box, and (x1, y1)
+            represents the position of the lower right. See Overview section for normalization.
```
```suggestion
            y1) format, where (x0, y0) corresponds to the position of the upper left corner in the bounding box, and (x1, y1)
            represents the position of the lower right corner. See Overview section for normalization.
```

sgugger commented on tests/test_modeling_layoutlm.py:
```diff
@@ -209,36 +235,266 @@ def test_for_masked_lm(self):
+
+def prepare_layoutlm_batch_inputs():
+    # Here we prepare a batch of 2 sequences to test a LayoutLM forward pass on:
+    input_ids = torch.tensor(
```
Can we either use a shorter inputs, or add some magic comment to make black ignore the long line, because this is a waste of vertical space :-)

sgugger commented on tests/test_modeling_layoutlm.py:
```diff
@@ -209,36 +235,266 @@ def test_for_masked_lm(self):
+    labels = torch.tensor(
+        [
+            [-100, 10, 10, 10, 9, 1, -100, 7, 7, -100, 7, 7, 4, 2, 5, 2, 8, 8, -100, -100, 5, 0, 3, 2, -100],
+            [
```
Same here.

sgugger reviewed (approved):
Thanks for all your work on this!

LysandreJik commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -709,12 +713,7 @@ class PreTrainedModel
-        output_type=BaseModelOutputWithPoolingAndCrossAttentions,
-        config_class=_CONFIG_FOR_DOC,
-    )
+    @replace_return_docstrings(output_type=BaseModelOutputWithPooling, config_class=_CONFIG_FOR_DOC)
```
```suggestion
    @replace_return_docstrings(output_type=BaseModelOutputWithPoolingAndCrossAttentions, config_class=_CONFIG_FOR_DOC)
```

LysandreJik commented on tests/test_modeling_layoutlm.py:
```diff
@@ -17,19 +17,26 @@
 
 
 class LayoutLMModelTester:
-    """You can also import this e.g from .test_modeling_bart import BartModelTester """
```
Good catch!

LysandreJik reviewed (approved):
LGTM, great job @NielsRogge! Thanks a lot for your contribution!

NielsRogge commented:
Thanks for the reviews, I've addressed all comments. There are 2 remaining things:

- in the code examples, I use both `tokenize()` and `convert_tokens_to_ids` as the bounding boxes (which are at word-level) need to be converted to token-level. Is there a better solution?
- according to @sgugger the input data on which the integration tests are ran are maybe too long, and black formatting causes them to be flattened vertically. Could you maybe fix this @LysandreJik?

LysandreJik commented:
I pushed the reformat you asked for @NielsRogge, make sure to pull before doing any more changes!

NielsRogge commented:
Ok thank you, so the only thing remaining is make the code examples more efficient? Is there a way to make the code block (see comment above) better?

patrickvonplaten commented on src/transformers/models/layoutlm/modeling_layoutlm.py:
```diff
@@ -738,21 +738,18 @@ def forward(
-            >>> bbox = torch.tensor(token_boxes).unsqueeze(0) # batch size of 1
-            >>> attention_mask = torch.tensor([1] * len(tokens)).unsqueeze(0) # batch size of 1
-            >>> token_type_ids = torch.tensor([0] * len(tokens)).unsqueeze(0) # batch size of 1
+            >>> encoding = tokenizer(' '.join(words), return_tensors="pt")
```
Thanks so much! That's much nicer
